packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,224 @@
+   */
+  match(route: string): RouteTreeNodeMetadata | null {
+    let bestMatchMetadata: RouteTreeNodeMetadata | null = null;
+    let lowestIndex = Infinity;
+
+    /**
+     * Recursively traverses the route tree to find the best match.
+     */
+    const traverse = (currentNode: RouteTreeNode, remainingSegments: string[]): void => {
+      const { insertionIndex, metadata, children } = currentNode;
+      if (!remainingSegments.length) {
+        // Reached a leaf node, check if it's a better match
+        if (metadata && insertionIndex < lowestIndex) {
+          bestMatchMetadata = metadata;
+          lowestIndex = insertionIndex;
+        }
+
+        return;
+      }
+
+      if (!children) {
+        return;
+      }
+
+      const [segment, ...restSegments] = remainingSegments;
+
+      // Try to match the exact segment
+      let childNode = children.get(segment);
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+
+      // Try to match with a wildcard
+      childNode = children.get('*');
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+    };
+
+    const segments = stripTrailingSlash(route).split('/');
+    traverse(this.root, segments);
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a nested object representation.
+   * @returns A nested object where each key is a route segment and each value is an object containing
+   *          metadata and possibly further nested route segments.
+   */
+  toObject(): RouteNodeObjectRepresentation {
+    const result: RouteNodeObjectRepresentation = {};
+
+    /**
+     * Recursively traverses the route tree to build the nested object.
+     */
+    const traverse = (
+      node: RouteTreeNode,
+      currentPath: string,
+      currentObj: RouteNodeObjectRepresentation,
+    ) => {
+      if (node.metadata) {
+        currentObj.metadata = node.metadata;
+      }
+
+      if (node.children) {
+        for (const [segment, childNode] of node.children) {
+          currentObj.children ??= {};
+          currentObj.children[segment] = {};
+          traverse(childNode, joinUrlParts(currentPath, segment), currentObj.children[segment]);
+        }
+      }
+    };
+
+    traverse(this.root, '', result);
+
+    return result;
+  }
+
+  /**
+   * Constructs a RouteTree from a nested object representation.
+   * @param nestedObject A nested object where each key is a route segment and each value contains
+   *                     metadata and possibly further nested route segments.
+   * @returns A RouteTree constructed from the nested object.
+   */
+  static fromObject(nestedObject: RouteNodeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    /**
+     * Recursively traverses the nested object to insert routes into the tree.
+     */
+    const traverse = (currentObj: RouteNodeObjectRepresentation, currentPath: string) => {
```

Comment:
Consider: You could also make a separate `walk` function for `RouteNodeObjectRepresentation` for better decoupling here too.

Reply:
Let me think about this following what we decide in https://github.com/angular/angular-cli/pull/28183#discussion_r1712408994